### PR TITLE
Allow users to call `study.optimize()` in multiple threads

### LINKED ISSUE #4049: Allow users to call `study.optimize()` in multiple threads.
### Motivation

For users who are familiar with `concurrent.futures`, I'd like to support the following usage that calls `study.optimize()` in multiple threads, although we can concurrently evaluate objective function in multiple threads via the `n_jobs` option of `study.optimize()`.

```python
import optuna
from concurrent.futures import ThreadPoolExecutor


def objective(trial):
    x = trial.suggest_float("x", -100, 100)
    y = trial.suggest_float("y", -100, 100)
    return x**2 + y


if __name__ == "__main__":
    study = optuna.create_study()
    with ThreadPoolExecutor(10) as pool:
        for i in range(10):
            pool.submit(study.optimize, objective, n_trials=10)

    print(f"Number of trials: {len(study.get_trials())}")
    print(f"Best params: {study.best_params}")
```

Also I'd like to support the following usage of `dask.distributed` (refs https://github.com/optuna/optuna/pull/2023#issuecomment-1268792200).

```python
import optuna

from dask.distributed import Client, wait


def objective(trial):
    x = trial.suggest_float("x", -100, 100)
    y = trial.suggest_float("y", -100, 100)
    return x**2 + y


if __name__ == "__main__":
    with Client("...") as client:
        print(f"Dask dashboard is available at {client.dashboard_link}")
        backend_storage = optuna.storages.InMemoryStorage()
        dask_storage = optuna.integration.DaskStorage(storage=backend_storage)
        study = optuna.create_study(direction="maximize", storage=dask_storage)

        futures = []
        for i in range(10):
            futures.append(client.submit(study.optimize, objective, n_trials=100, pure=False))
        wait(futures)

        print(f"Number of trials: {len(study.get_trials())}")
        print(f"Best params: {study.best_params}")
```

### Description

In the current implementation of Optuna, we can call `study.optimize()` in multiple processes like:

```python
import optuna
from concurrent.futures import ProcessPoolExecutor


def objective(trial):
    x = trial.suggest_float("x", -100, 100)
    y = trial.suggest_float("y", -100, 100)
    return x**2 + y


if __name__ == "__main__":
    storage = optuna.storages.RDBStorage(
        "sqlite:///example-db.sqlite3",
        engine_kwargs={"connect_args": {"timeout": 15.0}}
    )
    study = optuna.create_study(storage=storage)
    with ProcessPoolExecutor() as pool:
        for i in range(10):
            pool.submit(study.optimize, objective, n_trials=10)

    print(len(study.get_trials()))
```

However, it will be actually broken if we replace `ProcessPoolExecutor` with `ThreadPoolExecutor` due to following checks.

https://github.com/optuna/optuna/blob/6fc5392c458101d5bfda12d3f97dd88e11d4fd5b/optuna/study/_optimize.py#L53-L54
https://github.com/optuna/optuna/blob/6fc5392c458101d5bfda12d3f97dd88e11d4fd5b/optuna/study/_optimize.py#L121

As discussed at https://github.com/optuna/optuna/pull/2023#issuecomment-1268792200, the above checks are added at #491. In my recognition, it is originally intended to prevent the illegal calls of `study.optimize()`.
For example, Optuna's Trial object holds the Study object like [this](https://github.com/optuna/optuna/blob/master/optuna/trial/_trial.py#L26-L48), so components such as Sampler, Pruner, and the objective function can access the study through trials. It allows illegal calls from within the objective function like:

```python
def objective(trial):
    trial.study.optimize(objective, n_trials=10)
```

At least, I don't think it was intended to make the above use cases impossible. The above use cases will be fixed by simply removing the above three lines, but it may not be that easy as some features such as `study.stop()` and progress_bar may break.

### Alternatives (optional)

_No response_

### Additional context (optional)

_No response_

### PR DESCRIPTION
**Pull request by c-bata**
## Motivation

Close #4049

In #4051, I added a module-global thread local object by using the `contextvars` module. However, it contains some controversial changes. As an alternative approach, I opened this PR that just replaceing `threading.Lock` with `threading.local`.

## Description of the changes

Replace `threading.Lock` with `threading.local()`.

### CODE REVIEW

**Comment by c-bata** [comment]
@sile @hvy Could you review this PR if you have time?

**Review by hvy — APPROVED** [review]
Thanks, changes LGTM!
I left some minor comments or questions if you have time to take a look.

Also, we might want to consider testing more thoroughly if this multithreading support is to be documented or made part of the public API, but as for this current change, and to support the Dask storage/study seems good. 👍

File: optuna/study/_optimize.py [hunk 1]
```diff
@@ -134,6 +134,7 @@ def _optimize_sequential(
     time_start: Optional[datetime.datetime],
     progress_bar: Optional[pbar_module._ProgressBar],
 ) -> None:
+    study._thread_local.in_optimize_loop = True
```
[the remark this answers is not in the document]
Comment (c-bata) [reply optuna/study/_optimize.py:139]: > Might be good to leave a comment this it's set to True in _optimize_sequential and not _optimize. Not a strong opinion.

Sounds good. I pushed 7a2bf82f5.

> Do you think adding a test for this? That is, try nested optimizations with n_jobs > 1 is overkill? 🤔

There are three test cases that will be broken if this logic is moved to `_optimize`. So I'd say further test cases can be omitted.

```diff
$ git diff
diff --git a/optuna/study/_optimize.py b/optuna/study/_optimize.py
index 7b2d897cb..aaaf639ce 100644
--- a/optuna/study/_optimize.py
+++ b/optuna/study/_optimize.py
@@ -60,6 +60,7 @@ def _optimize(
     progress_bar = pbar_module._ProgressBar(show_progress_bar, n_trials, timeout)

     study._stop_flag = False
+    study._thread_local.in_optimize_loop = True

     try:
         if n_jobs == 1:
@@ -134,7 +135,6 @@ def _optimize_sequential(
     time_start: Optional[datetime.datetime],
     progress_bar: Optional[pbar_module._ProgressBar],
 ) -> None:
-    study._thread_local.in_optimize_loop = True
     if reseed_sampler_rng:
         study.sampler.reseed_rng()
```

```
$ pytest tests/study_tests/
...
================================================================================ short test summary info =================================================================================
FAILED tests/study_tests/test_study.py::test_stop_n_jobs - RuntimeError: `Study.stop` is supposed to be invoked inside an objective function or a callback.
FAILED tests/study_tests/test_study.py::test_optimize_progbar_no_constraints[2] - RuntimeError: `Study.stop` is supposed to be invoked inside an objective function or a callback.
FAILED tests/study_tests/test_study.py::test_optimize_without_progbar_no_constraints[2] - RuntimeError: `Study.stop` is supposed to be invoked inside an objective function or a callback.
======================================================================== 3 failed, 591 passed in 80.39s (0:01:20) ========================================================================
```

**Review by sile — APPROVED** [review]
Sorry for the late review. This PR looks perfect!

**Comment by hvy** [comment]
Thanks for the swift alternative @c-bata and @sile for the review. Let me merge this PR since it has 2 approvals now.

**Pull request merged**